src/front/wgsl/mod.rs

```diff
@@ -466,6 +490,126 @@ impl<'a> Error<'a> {
     }
 }
 
+/// Represents the type of the constructor
+///
+/// Vectors, Matrices and Arrays can have partial type information
+/// which later gets inferred from the constructor parameters
+enum ConstructorType {
+    Scalar {
+        kind: crate::ScalarKind,
+        width: Bytes,
+    },
+    PartialVector {
+        size: crate::VectorSize,
+    },
+    Vector {
+        size: crate::VectorSize,
+        kind: crate::ScalarKind,
+        width: Bytes,
+    },
+    PartialMatrix {
+        columns: crate::VectorSize,
+        rows: crate::VectorSize,
+    },
+    Matrix {
+        columns: crate::VectorSize,
+        rows: crate::VectorSize,
+        width: Bytes,
+    },
+    PartialArray,
+    Array {
+        base: Handle<crate::Type>,
+        size: crate::ArraySize,
+        stride: u32,
+    },
+    Struct(Handle<crate::Type>),
+}
+
+impl TryFrom<ConstructorType> for TypeResolution {
```

Comment:
is `TryFrom` as a trait really necessary here? Could it instead be just a method on `ConstructorType`?

Reply:
Looking at it again I think a method returning an Option would be better.